### PR TITLE
vscode: do not autoformat markdown or toml files

### PR DESCRIPTION
**Pull request by thoughtpolice**
Summary: This caught me several times and several instances of autoformatting have slipped by in some of my commits; generally this is a PITA when you're doing something like modifying markdown files and Visual Studio Code "helpfully" autoformats the entire document for a 1-byte change, creating a 100+ line diff out of thin air.

This is of course a function of the specified users autosave configuration and enabled extensions. I can't possibly account for everyone's integrations or workflows, but at least for me, I think, it's OK to turn off autoformatting here because we aren't enforcing a consistent one anyway.

I do think that adding autoformatting for some of these things *may* be a good idea. But in the mean time I think this is just a helpful safeguard.

As an example of this problem, I've included a small diff to remove a bunch of redundant whitespace from the top-level Cargo workspace file; this is exactly the kind of autoformatting that I think isn't good unless everyone agrees on it and it only serves to bloat diffs if someone changes it later (and no, there isn't really a standardized format for Cargo.toml yet from what I can tell — though, tellingly, the related style guides for Cargo.toml in the manual don't seem to recommend things like "add 60 characters of whitespace for no reason to align comments to column 80".)

Change-Id: Iea17b611d58e65135fdc702f7b59f46c

# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

**Comment by arxanas** [comment]
Works for me 🤷

**Comment by joyously** [comment]
Does the project have a .editorconfig file?
See https://editorconfig.org/

**Comment by thoughtpolice** [comment]
> Does the project have a .editorconfig file? See https://editorconfig.org/

I actually checked `.editorconfig` first before anything else but, alas, it has no keys or options for controlling things like "format this file on save." It's mostly just basic character encoding and whitespace options, still.

(While I like `.editorconfig` a bunch, I think part of its appeal is that it's simple and very easy to semantically understand. On the other hand I think that means it's hard to evolve, and so it's pretty much "done" and so some of these things like "format on save" are things it missed out on specifying.)

**Comment by ilyagr** [comment]
I'm not a huge fan of this: if somebody set up autoformatting on save that actually works for them, why shouldn't it work in our repo? If somebody set up autoformatting on save that *doesn't* work for them, why should we prevent them from learning that they don't like they config?

Either way, since this config doesn't make autoformatting work perfectly for our repo, I don't think it's worth making somebody's editor behave differently in our repo compared to another repo.

For TOML, formatting-on-save seems less important, so my feelings are less strong either way. It's also not a hill I'd die on; if this really makes people's lives more convenient, I'm OK with it.

As an aside, we'd benefit from a good Markdown auto-formatter; I'd rather standardize on some plugin (is ["Prettier"](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) good? I think it would work in other editors), set up a presubmit check, and use `.vscode` to recommend it and set up a config. That'd require some research in whether we actually like its formatting, whether it's a reasonable thing to install in editors other than VS Code, and how to set up a presubmit check.

BTW, I set up formatting on save for `rs` files only (mainly because it matches the presubmit check). For Markdown, I use the ["Rewrap"](https://marketplace.visualstudio.com/items?itemName=stkb.rewrap) plugin currently, which doesn't do the entire job but seems good at what it does do.

**Comment by martinvonz** [comment]
> I'd also be really happy with this, but I'll be honest, part of the reason I submitted this instead of that is because formatting discussions can get into wonderfully long bikesheds. :)

I personally care much more that we have *some* standardized formatting than exactly what it looks like.

**Comment by martinvonz** [comment]
Ah, I thought you meant bikeshedding the details of how to format. I agree that don't require a particular editor.

**Comment by ilyagr** [comment]
I looked into Markdown formatting a bit. I might try formatting the jj repo with the two options I found when I get a moment.

I also don't want to create any barriers for first-time contributors that want to help us with documentation. Unfortunately, there seems to be no option that's trivial to install in the Rust ecosystem.

So, I think any presubmit checks for Markdown formatting would have to be optional.

### Prettier

Prettier seems to be pretty bad, at a glance. It converts `*blah*` to `_blah_`, which I find annoying. Less definitely but perhaps more importantly, Markdown support seems to be an afterthought for them, it's more of a HTML/CSS/JS/etc tool. See also https://github.com/prettier/prettier/issues/11828 for problems people have with it.

As a pro, it's very well integrated in different editors. It can also be run with https://pre-commit.com/ (see below).


### Mdformat
There's https://github.com/executablebooks/mdformat. (This is different than an internal Google tool of the same name). It claims to be better than Prettier, but doesn't have any editor integrations I could find. It only integrates with https://pre-commit.com/.

`mdformat` also makes no effort to be easy to use outside the Python ecosystem. https://pre-commit.com/ seems to be a little better about that, and claims to be able to install `mdformat`.

Since jj doesn't support pre-commit hooks, we'd probably have to run it with https://pre-commit.com/ manually (`pre-commit run`). https://pre-commit.com/ does have some editor integrations, but they seemed a little sketchy.

#### Aside on pre-commit hooks
I'm not sure whether pre-commit hooks designed for `git` are a good idea for `jj`, as they might be too slow. They budget latency for a tool that's run on explicit user action, but `jj` would run them on every save. A command to run `pre-commit run` with something like [`jj run`](https://github.com/martinvonz/jj/issues/1869) would probably be better.

After writing this, I realized there's https://github.com/martinvonz/jj/issues/405 tracking this. I think it came to a similar conclusion.


### Aside on Chromium
FWIW, the Chromium project seems to have [given up on formatting markdown](https://chromium.googlesource.com/chromium/src/+/main/styleguide/markdown/markdown.md#:~:text=Unfortunately%2C%20git%20cl%20format%20will%20NOT%20do%20this%20for%20you.).

**Comment by ilyagr** [comment]
You can take a look at my first attempt at auto-formatting: https://github.com/ilyagr/jj/commit/mdformat and https://github.com/ilyagr/jj/commit/prettier.

Clearly, we'd need to exclude some files from overly aggressive formatting, but this can give a taste of what it looks like. Also, you can try the `pre-commit run --all-files` workflow and see if it works for you.

Prettier also formats a bunch of non-md files. Pre-commit should be able to prevent that, I think.

**Comment by ilyagr** [comment]
I found another interesting Markdown formatter: https://github.com/dprint/dprint. It's used by the `comprehensive-rust` course: https://github.com/google/comprehensive-rust/commit/aaca44f62b8e562cef5044c4f61ff20e45b97c73. Apparently, it can both format Markdown and call `cargo fmt`.

**Comment by thoughtpolice** [comment]
While working on https://github.com/martinvonz/jj/pull/2273 I was once again bitten by vscode "helpfully" doing autoformatting due to my settings using it by default, so to stop the diff from bloating, I had to snip out the relevant partsput it in another file, `jj restore` the original, then insert the changes back into the file. Rather annoying.

Regardless of what we choose, I realize I'm not sure what kind of editor integration it will end up having, so I'd *still* like to see this merged. For example `dprint` looks good, but if there's no vscode support, I'd prefer if we used it as a CI check and disabled the users editor integration. Otherwise your editor is basically guaranteed to break the standard. If vscode support came along later, we could change the preferences to use it for formatting those files.

**Review by martinvonz — APPROVED** [review]
LGTM. I only skimmed the Conversion tab but I didn't see any objections from anyone else either.

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+    // NOTE (aseipp): don't autoformat {markdown, toml, et cetera} files, since
+    // there's no real standardized format we've ever tried to agree on; until
+    // then, such a feature tends to make diffs much worse and various people's
+    // editors config may have some "helpful" plugins that autoformat these
+    // filetypes, so this disables a footgun for everyone.
+    "[markdown]": {
```
Comment (ilyagr) [inline .vscode/settings.json:7]: I'm not very happy about this. For one, this is incompatible with somebody having another per-workspace config for their `jj` repo.

I wonder whether VSCode provides some sort of "suggested config", so this'd work similarly to the `envrc.recommended`.

File: .vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+        "editor.formatOnSave": false
+    },
+    "[toml]": {
+        "editor.formatOnSave": false
+    }
+}
```
Comment (ilyagr) [inline .vscode/settings.json:13]: My preferred (but imperfect) solution to such a usecase would be to have you put .vscode into a global gitignore or .git/exclude and have your own vscode config.

On problem I see with this is that finding the per-repo `.git/exclude` with jj is tricky with non-colocated repos (I'm not even sure it's supported).

File: .vscode/settings.json (hunk 1 above)
Comment (ilyagr) [reply .vscode/settings.json:7]: I'm not sure, BTW, whether this problem is theoretical or practical: do people often use `.vscode/settings.json` for something else? So, if this matters to you enough (and the sort-of-solution from my other comment doesn't work), we could still go with it.

**Comment by thoughtpolice** [comment]
Closing this for now, as there's division on this and it can be worked around locally for me in various ways.

**Pull request closed without merging**